### PR TITLE
Speech filtration implants pt. 2: coding boogaloo

### PR DESCRIPTION
**Pull request by MACIEKBAKI**
## About The Pull Request

Doktor, turn off my n word inhibitor!
## Changelog
:cl:
add: Speech Filtration Implants
/:cl:

### CODE REVIEW

**Comment by MACIEKBAKI** [comment]
apologies for the autistic commit history github refused to work with me

File: code/modules/mob/living/living_say.dm [hunk 1]
```diff
@@ -92,6 +92,12 @@ GLOBAL_LIST_INIT(message_modes_stat_limits, list(
 
 	return new_msg
 
+// This is purely for the word inhibitor implant
+// Overwritten in /mob/living/carbon for the implant logic
+// ((mob/living has no organs))
+/mob/living/proc/midflight_check(message) //todo: come up with a better name for proc
```
[the remark this answers is not in the document]
Comment (MACIEKBAKI) [reply code/modules/mob/living/living_say.dm:98]: oh yeah that's a good idea thanks mate

File: code/modules/surgery/organs/cybernetics/augments_internal_wordblock.dm [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+	return mgr
+
+/obj/item/organ/cyberimp/brain/wordblocker/proc/handle_speech(datum/source, list/speech_args)
+	SIGNAL_HANDLER
+
+	if(!mgr)
+		return
+
+	var/newmsg = mgr.process_message(speech_args[SPEECH_MESSAGE])
+	if(owner && (!newmsg || newmsg == FALSE || newmsg == ""))
```
Comment (TheChosenEvilOne) [inline code/modules/surgery/organs/cybernetics/augments_internal_wordblock.dm:54]: IIRC an empty string is false in byond.
```suggestion
	if(owner && !newmsg)
```

File: code/modules/surgery/organs/cybernetics/augments_internal_wordblock.dm [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+	if(!active)
+		return message // don't process the string
+	var/regex/r = new(blocked_word, (case_sensitive ? "gm" : "gmi"))
+	// regex can be replaced with replacetext() if it gets abused or something
+	var/t = message
+	if(replace)
+		t = r.Replace(message, replace_phrase)
+	else
+		var/find = r.Find(message)
+		if(find != 0 && (find+1) < length(message)) // block "lean" = "I love lean, you should too" -> "i love l-"
```
Comment (TheChosenEvilOne) [inline code/modules/surgery/organs/cybernetics/augments_internal_wordblock.dm:107]: not really necessary, but looks better IMO.
```suggestion
		if(find > 0 && (find + 1) < length(message)) // block "lean" = "I love lean, you should too" -> "i love l-"
```

**Review by TheChosenEvilOne — APPROVED** [review]
seems good, I didn't look at the UI because I hate tgui.

File: code/modules/surgery/organs/cybernetics/augments_internal_wordblock.dm [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	if(mgr)
+		. += "<span class='info'>The implant currently does [mgr.visible ? "" : "not"] allow remote access. Use a screwdriver to change this.</span>"
+
+/obj/item/organ/cyberimp/brain/wordblocker/screwdriver_act(mob/living/user, obj/item/I)
+	. = ..()
+	if(.)
+		return TRUE
+	I.play_tool_sound(src)
+	if(!mgr || mgr == null)
+		to_chat(user, "<span class='notice'>You try to reconfigure [src], but your [I] phases straight through it. You should probably notify a coder. </span>")
```
Comment (stanalbatross) [inline code/modules/surgery/organs/cybernetics/augments_internal_wordblock.dm:74]:
```suggestion
		to_chat(user, "<span class='notice'>You try to reconfigure \the [src], but your [I] phases straight through it. You should probably notify a coder. </span>")
```

File: code/modules/surgery/organs/cybernetics/augments_internal_wordblock.dm [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+		return
+	if(locked)
+		to_chat(user, "<span class='notice'>You try to reconfigure [src], only to find it's remote access switch glued in the [mgr.visible ? "ON" : "OFF"] position.</span>")
```
Comment (stanalbatross) [inline code/modules/surgery/organs/cybernetics/augments_internal_wordblock.dm:77]:
```suggestion
		to_chat(user, "<span class='notice'>You try to reconfigure \the [src], only to find its remote access switch glued in the [mgr.visible ? "ON" : "OFF"] position.</span>")
```
laugh at the ESL.

File: code/modules/surgery/organs/cybernetics/augments_internal_wordblock.dm [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+		return
+	mgr.visible = !mgr.visible
+	to_chat(user, "<span class='notice'>You reconfigure [src] to [mgr.visible ? "" : "not"] allow remote access.</span>")
```
Comment (stanalbatross) [inline code/modules/surgery/organs/cybernetics/augments_internal_wordblock.dm:80]:
```suggestion
	to_chat(user, "<span class='notice'>You reconfigure \the [src] to [mgr.visible ? "" : "not"] allow remote access.</span>")
```

File: code/modules/surgery/organs/cybernetics/augments_internal_wordblock.dm [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+		return
+	speech_args[SPEECH_MESSAGE] = newmsg
+
+
+
+/obj/item/organ/cyberimp/brain/wordblocker/examine(mob/user)
+	. = ..()
+	. += "<span class='info'>A little screen reads: 'Implant Unique ID: [id]'</span>"
+	if(mgr)
+		. += "<span class='info'>The implant currently does [mgr.visible ? "" : "not"] allow remote access. Use a screwdriver to change this.</span>"
```
Comment (stanalbatross) [inline code/modules/surgery/organs/cybernetics/augments_internal_wordblock.dm:66]:
```suggestion
		. += "<span class='info'>The implant currently [mgr.visible ? "allows" : "does not allow"] remote access. Use a screwdriver to change this.</span>"
```
laugh at the ESL part two

**Review by stanalbatross — CHANGES_REQUESTED** [review]

File: code/modules/surgery/organs/cybernetics/augments_internal_wordblock.dm (hunk 7 above)
Comment (TheChosenEvilOne) [reply code/modules/surgery/organs/cybernetics/augments_internal_wordblock.dm:66]: Both are correct:
"[it] does allow remote access" and "[it] does not allow remote access"
"[it] allows remote access" and "[it] does not allow remote access"
the word "does" in the former adds emphasis to the verb "allow".

**Comment by QuakeIV** [comment]
im going to try to review this one more thoroughly to evaluate the quality of the actual code, but i am personally very much in favor of this idea

**Comment by QuakeIV** [comment]
> apologies for the autistic commit history github refused to work with me

i want you to squash these commits before you merge i can show you how to do this if you cant figure it out

**Comment by MACIEKBAKI** [comment]
> im going to try to review this one more thoroughly to evaluate the quality of the actual code, but i am personally very much in favor of this idea

Why are you in favor, if I might ask?

File: code/modules/surgery/organs/cybernetics/augments_internal_wordblock.dm (hunk 7 above)
Comment (MACIEKBAKI) [reply code/modules/surgery/organs/cybernetics/augments_internal_wordblock.dm:66]: *fixes sentence that was already correct*
"laugh at ESL"

**Comment by QuakeIV** [comment]
> 

because its hilarious

**Comment by QuakeIV** [comment]
yea im going to stop asking people to squash, it turns out we do that for you anyways as standard practice so i need to just get used to that

**Comment by QuakeIV** [comment]
show this working please that is the last roadblock for this i think

OBS + upload to youtube, shouldnt be too hard

**Comment by MACIEKBAKI** [comment]
https://www.youtube.com/watch?v=9Cg5dRa_9Wc

**Comment by QuakeIV** [comment]
sweet

**Comment by QuakeIV** [comment]
approved and merged (hey look at that we actually test merged and talked about it instead of being an absolute n-)

**Pull request merged**